src/System.Security.Cryptography.X509Certificates/src/System/Security/Cryptography/X509Certificates/TbsCertificate.cs

```diff
@@ -1,194 +0,0 @@
-                //   30 [length]
-                //     First Extension
-                //     Second Extension
-                //     ...
-
-                // An interesting quirk of skipping null values here is that
-                // Extensions.Count == 0 => no extensions
-                // Extensions.ContainsOnly(null) => empty extensions list
-
-                HashSet<string> usedOids = new HashSet<string>(Extensions.Count);
-
-                foreach (X509Extension extension in Extensions)
-                {
-                    if (extension == null)
-                        continue;
-
-                    if (!usedOids.Add(extension.Oid.Value))
-                    {
-                        throw new InvalidOperationException(
-                            SR.Format(SR.Cryptography_CertReq_DuplicateExtension, extension.Oid.Value));
-                    }
-
-                    encodedExtensions.Add(extension.SegmentedEncodedX509Extension());
-                }
-
-                byte[][] extensionField = DerEncoder.ConstructSegmentedSequence(
-                    DerEncoder.ConstructSegmentedSequence(encodedExtensions));
-                extensionField[0][0] = DerSequenceReader.ContextSpecificConstructedTag3;
-
-                encodedFields.Add(extensionField);
-            }
-
-            return DerEncoder.ConstructSequence(encodedFields);
-        }
-
-        private static byte[][] EncodeValidityField(DateTimeOffset validityField, string propertyName)
-        {
-            /* https://tools.ietf.org/html/rfc3280#section-4.1.2.5
-             4.1.2.5  Validity
-
-                The certificate validity period is the time interval during which the
-                CA warrants that it will maintain information about the status of the
-                certificate.  The field is represented as a SEQUENCE of two dates:
-                the date on which the certificate validity period begins (notBefore)
-                and the date on which the certificate validity period ends
-                (notAfter).  Both notBefore and notAfter may be encoded as UTCTime or
-                GeneralizedTime.
-
-                CAs conforming to this profile MUST always encode certificate
-                validity dates through the year 2049 as UTCTime; certificate validity
-                dates in 2050 or later MUST be encoded as GeneralizedTime.
-
-                The validity period for a certificate is the period of time from
-                notBefore through notAfter, inclusive.
-            */
-
-            DateTime utcValue = validityField.UtcDateTime;
-
-            // On the one hand, GeneralizedTime easily goes back to 1000, and possibly to 0000;
-            // but on the other, dates before computers are just a bit beyond the pale.
-            if (utcValue.Year < 1950)
-            {
-                throw new ArgumentOutOfRangeException(propertyName, utcValue, SR.Cryptography_CertReq_DateTooOld);
-            }
```

Comment:
Was this exceptional state lost?  Or is it a repeat of non-modified code?

Reply:
It was lost, now it allows years below 1950 and relies on the checks in AsnWriter to do the right thing.

Reply:
...but it looks like AsnWriter doesn't do checks for negative years.

T-REC-X.680-201508 sec 46 says that GeneralizedTime should be written using "ISO 8601, 4.1.2.2 – Basic format". It doesn't specify any additional restrictions. The 4.1.2.2 representation doesn't allow for negative years as far as I can tell (the expanded representation has to be used). I guess `WriteGeneralizedTimeCore` method should be modified to reject years < 0 just as it rejects years > 9999.

Reply:
Ah, it doesn't need to check for negative years since DateTime can't represent them.